Fix multiple harvester dock sequence bugs

reaperrr opened the pull request:
Fixes
- drag activity running even if refinery died while turning to dock angle
- dock anim running even if refinery died while dragging towards it
- undock anim playing even if dock anim didn't run for the above or some other reason
- and some more (see commit messages)

Fixes #18422.
Fixes #18733.

pchote commented:
> Activities/HarvesterDockSequence.cs(73,1): error SA1028: Code should not contain trailing whitespace [/home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/OpenRA.Mods.Common.csproj]

pchote commented:
My initial thought was that the code looks sensible, but when I test this by adding `Refinery.Kill(self);` in the `DockingState.Turn` case I find that this doesn't seem to work!

Are you able to reproduce this using the same testcase?

pchote commented on OpenRA.Mods.Common/Activities/HarvesterDockSequence.cs:
```diff
@@ -62,10 +62,15 @@ public override bool Tick(Actor self)
+						OnStateDock(self);
+					}
+					else
+						dockingState = DockingState.Undock;
```
Changing this to to `DockingState.Complete` improves the situation, but may have other side effects?

reaperrr replied:
`Undock` is correct, as mentioned below I found a proper fix for this.

pchote commented:
We can fix the harvester dragging in and out if the refinery was killed while turning by moving the drag activity from DockingState.Turn to its own case:
```csharp
case DockingState.Drag:
	if (!Refinery.IsInWorld || Refinery.IsDead)
		return true;

	dockingState = DockingState.Dock;
	if (IsDragRequired)
		QueueChild(new Drag(self, StartDrag, EndDrag, DragLength));

	return false;
```

Note that these suggestions don't help with the second glitch in that issue, which is easily reproduced by issuing a move order while the harvester is turning to dock.

reaperrr committed:
Fix dock animation continuing after refinery died

OnStateDock would always be triggered, even if the
refinery was killed before the harvester would finish
docking.

reaperrr committed:
Fix undock anim playing even if docking was cancelled

...before dock anim could run.

The undock animation would play even if the dock anim
hadn't run (meaning the sequence cancelled before the docking
completed, for example due to refinery death).

reaperrr commented:
> My initial thought was that the code looks sensible, but when I test this by adding Refinery.Kill(self); in the DockingState.Turn case I find that this doesn't seem to work!
> Are you able to reproduce this using the same testcase?

The code wasn't wrong, it's just that the issue basically consists of multiple bugs.

The problem was that the undock anim would still play on "undock" even if the dock anim hadn't. Fixed that, but I pushed before seeing the above additional suggestions so I'll update this again.

reaperrr committed:
Fix harvs dragging to refinery even if it died while turning

reaperrr commented:
Updated.

pchote commented:
I can still trigger a glitch by issuing a move order while the TD harvester is turning. I thought that adding `IsCancelling` to the check at the start of `DockingState.Drag` would have solved this, but seems not.

reaperrr committed:
Fix harv dock sequence issuing drag despite cancellation

reaperrr commented:
> I can still trigger a glitch by issuing a move order while the TD harvester is turning. I thought that adding IsCancelling to the check at the start of DockingState.Drag would have solved this, but seems not.

Updated, should be fixed now (last commit).

pchote commented:
Confirmed that this does not fix #18733. Repro case: add `Refinery.Kill(self);` to the start of `case DockingState.Undock:`.

Do you want to fix that here too?

pchote commented:
Code changes look good to me and i've confirmed that all the other death cases work correctly.

reaperrr committed:
Fix TS harvesters getting stuck in unload graphics

when refinery is destroyed while unloading.

reaperrr commented:
Updated.

> Do you want to fix that here too?

Should be fixed now, if my eyes didn't deceive me in my quick test.

abcdefg30 commented on OpenRA.Mods.Common/Activities/HarvesterDockSequence.cs:
```diff
@@ -85,7 +102,7 @@ public override bool Tick(Actor self)
 					Harv.LastLinkedProc = Harv.LinkedProc;
 					Harv.LinkProc(self, null);
-					if (IsDragRequired)
+					if (IsDragRequired && wasDraggedToDock)
```
`wasDraggedToDock` seems unnecessary since `IsDragRequired` can't change, right?

reaperrr replied:
The idea is to only play the reverse drag if the initial drag activity also ran, otherwise the actor will insta-jump to dragEnd and reverse-drag from there.

pchote replied:
As an alternative (not sure its better) we could make `IsDragRequired` not readonly and set it to false if the refinery is killed.

abcdefg30 replied:
If the initial drag activity doesn't run we should never reach this point, shouldn't we?

abcdefg30 replied:
https://github.com/OpenRA/OpenRA/pull/18737/files#diff-b3d67bffe7672e41778dd47afabf1ea99403a2542bf121c5ff4eb4cc82bf6904R64 cancels without going to another state.

reaperrr replied:
`Cancel()` unconditionally setting `dockingState = DockingState.Undock;` is the problem here. We should probably just drop that and add `IsCanceling` checks where necessary.

abcdefg30 commented on OpenRA.Mods.Cnc/Activities/VoxelHarvesterDockSequence.cs:
```diff
@@ -45,10 +45,15 @@ public override void OnStateDock(Actor self)
+			if (!body.Docked)
+			{
+				dockingState = DockingState.Complete;
+				return;
```
The `return` statement here is superfluous since there is no code behind it anyway. (The rest of the method are in `else` branches.)

abcdefg30 reviewed (commented):
Dead actors are never in the world, so I'm not sure if it makes sense here to check both. (Just `IsInWorld` should be enough.)
On that note, do we have some sort of guard against a refinery just getting captured between those states? (Similarly to how we now check if it got removed from the world in-between.)

reaperrr commented:
> Dead actors are never in the world, so I'm not sure if it makes sense here to check both. (Just `IsInWorld` should be enough.)

`IsDead` returns `true` as soon as health drops to zero, while `Actor.Dispose()` only sets `IsInWorld` to `false` in a frame end task. I'm a bit wary of not checking `IsDead`, though admittedly I might be a bit too paranoid here.

pchote commented:
There are a bunch of places across the code that check both IsDead and IsInWorld. How about we keep it here and, if/when we're confident that just IsInWorld is the correct change, have a new PR that fixes this everywhere?

pchote commented:
Do we have a plan on how to proceed here?

reaperrr commented:
> Do we have a plan on how to proceed here?

While implementing
> We should probably just drop that and add IsCanceling checks where necessary.

I stumbled over another design flaw of the dock sequence, namely that `DockingState.Complete` still unconditionally called `Undocked` on the refinery if it's alive, as well as drag. Since I was recently swamped with work and stuff, I didn't have the mental bandwidth to finish this, but I'm working on it right now and will update soon.

reaperrr committed:
Fix dock sequence dock/undock notifications

- move Refinery dock/undock notifications to OnStateDock/-Undock
and only call Undock if Dock was also called.
- Add INotifyHarvesterAction support to VoxelHarvesterDockSequence.

abcdefg30 reviewed (commented):
> On that note, do we have some sort of guard against a refinery just getting captured between those states?

With the right timing I can still capture a refinery right before the harvester starts unloading. It doesn't cancel the activity but proceeds to unload its cargo into the banks of the enemy who just captured the refinery (while remaining in the player's control).

pchote commented:
That's quite a different bug (and IMO a rare enough edge case that it may not be worth fixing at all) to the issues being fixed here. Can we track and fix that in a separate issue/PR?

abcdefg30 commented:
Fair enough, I thought we wanted to tackle all of the while we're here. Done at #18835.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ed9151d2acaf6fdacb2273198b447d846b3509f8...94de7eab70433a8d152c95f2dfa056f526f8f5e2)
